### ISSUE TITLE
Add support for OpenDoas for privilege escalation

### ISSUE THREAD

**Issue by ComplicatedPenguin** [veracrypt/VeraCrypt#823]
# Desired behavior
OpenDoas support for privilege escalation

# Additional information
I was trying to encrypt an external drive using VeraCrypt and i started VeraCrypt volume creation wizard, and I reached the part that I must provide password for my new encrypted VeraCrypt volume and I wrote my password and after that VeraCrypt showed an error about sudo not being available:
![veracrypt](https://user-images.githubusercontent.com/85984161/132102383-e8e8ade5-7be7-48b1-8f34-a68dc16d9c4b.jpg)
One possible way I suggest this may be done is implementing a checker that checks for sudo and if it  finds out host system doesn't have sudo installed tries to run the same task with doas.
[OpenDoas GitHub repository](https://github.com/Duncaen/OpenDoas)
# Your Environment

**VeraCrypt version:** 1.24-Update7-1

**Operating system and version:** Arch Linux

**System type:** 64-bit

**Comment by 766F6964** [comment]
I agree with this, it would be nice if veracrypt does not depend on sudo, - but instead can also be used with e.g. doas.

**Comment by CreamySweets** [comment]
I agree this is a nice feature to have.
As an addition, symlinking sudo to doas tries to work but fails, stating "_sudo: invalid option -- 'S'_"

**Comment by xioren** [comment]
Yes please. I just switched to doas from sudo and to my dismay I can no longer use veracrypt.

**Comment by idrassi** [comment]
I have started doing changes to support `doas` but the blocking issue is that it doesn't support reading the password from pipe. This is needed so we can send the password from VeraCrypt to the forked `doas` performing privileged operation. 

I have found that someone who needed the same approach opened an issue on OpenDoas but it was closed because it will not be implemented (https://github.com/Duncaen/OpenDoas/issues/33). 

So it seems that it will not be possible to support `doas` in VeraCrypt in the near future. Of course, I'm open to ideas/proposals for possible workarounds.

**Comment by xioren** [comment]
Well that's a bummer. Thanks for trying.

**Comment by 6e6f61** [comment]
Running `veracrypt` as root is almost a work around, but the mount will be owned by it too :/

**Comment by soracqt** [comment]
I found workaround can be decrypt  without root.

```bash
if [ ! -f /etc/udisks2/tcrypt.conf ] ; then sudo touch /etc/udisks2/tcrypt.conf; fi
```

and reboot . then thunar will be able to decrypt the veracrypt volume.

Otherwise, option to use udisksctl from cui

```bash
udiskctl unlock -b /dev/sdX#
```

## Mounting veracrypt volume

```bash
udisksctl mount -b /dev/dm-#
```

**Comment by TheDuchy** [comment]
Would be solved by https://github.com/veracrypt/VeraCrypt/issues/887

**Comment by withs** [comment]
> Running `veracrypt` as root is almost a work around, but the mount will be owned by it too :/

maybe this can be a workaround.
as i mentioned in this comment, i've only tested it on openbsd 
https://github.com/veracrypt/VeraCrypt/issues/589#issuecomment-1872183317

**Comment by idrassi** [comment]
I revisited this issue and the situation has not fundamentally changed since my previous comment.

On Linux, VeraCrypt currently starts an elevated `veracrypt --core-service` using `sudo -S` and then communicates with that elevated process over pipes. This requires a privilege-escalation tool that can receive the authentication password from VeraCrypt through stdin. OpenDoas does not provide an equivalent mechanism, and the upstream request for this behavior was closed without implementation, so `doas` cannot be used as a transparent replacement for the current `sudo` flow.

The only form of OpenDoas support that looks technically feasible is limited non-interactive support: if the user has an explicit `nopass` rule in `/etc/doas.conf`, VeraCrypt could potentially detect/use `doas -n` as a fallback when `sudo` is unavailable. This would not cover the normal case where VeraCrypt asks the user for an administrator password, because that password cannot be passed to `doas`.

For broader support on systems without sudo, the better direction is probably not OpenDoas-specific password handling, but a separate elevation backend such as polkit/pkexec or an installed privileged helper. That is a larger design change and overlaps with #887.

So my current position is:
- full OpenDoas support is still blocked by OpenDoas design/limitations
- limited `doas -n`/`nopass` support could be considered if someone contributes a clean patch
- general non-sudo elevation should be handled separately, likely through polkit or a dedicated helper
